librz/include/rz_asm.h

```diff
@@ -141,6 +141,7 @@ typedef struct rz_asm_plugin_t {
 	char *(*mnemonics)(RzAsm *a, int id, bool json);
 	const char *features;
 	const char *platforms;
+	RzAsmTokenPattern token_patterns[RZ_ASM_NUM_TOKEN_PATTERNS]; //< All asm token patterns, ordered by priority. Highest priority at index 0.
```

Comment:
RzPVector? but yes, a pointer is better

Reply:
Pointer is fine for me as well :+1:

Reply:
The important point is api and abi compatibility. I think a callback like `char *(*token_pattern)(RzAsmTokenType type)` (returning just the pattern because all other members of `RzAsmTokenPattern` are then implicit or generated) can work best.

Reply:
> The important point is api and abi compatibility.
Sorry if I missed the docs about this. But what do you mean with API compatibility?

> `char *(*token_pattern)(RzAsmTokenType type)`

What about https://github.com/rizinorg/rizin/pull/2417/commits/e311159e7783a3b29ab0353f8c77d0da53280e65 ?

Reply:
RzList is fine. If we don't want to rely on RzList we can just use a simple array terminated by a NULL element.

Reply:
> Sometimes there are more patterns assigned to the same type (see code). So a `type <-> pattern` one to one match is not always possible.

Would `(<pattern0>)|(<pattern1>)|...` as part of the regex work too for you?

Reply:
No. The patterns need a priority because the same symbols can occur in different types.
E.g. mnemonics can contain digits (`trap0`) and numbers can contain letters (`0xff, 0b10` etc.) and so on.

Since there can be so many different mnemonics, mnemonic regex should match any kind of alphanumeric string (`[[:alnum:]]+`) with the exception of .

Lets assume we have the asm string `trap0(0xff) ; R13 = memw(0x5000 + R13in)`. We have to:
1. Match hex numbers  - `type = number`
2. Match registers - `type = register`
3. Match mnemonics with number (`trap0`). Necessary, otherwise the `0` would be matched by the decimal numbers - `type = mnemonic`
4. Match decimal numbers - `type = number`
5. Match other mnemonics (`memw`) - `type = mnemonic`
6. Match separators - `type = separator`

Simply ORing of regex patterns would unfortunately not work here.
Also think of more complex asm strings like dex. Maybe an actual parser with a grammar for each arch is better?

Reply:
I see. For such an approach, returning a list from a callback sounds good. A context free grammar is probably strong enough for most archs too. mpc comes to mind as a solution to implement this, but we want to get rid of that dependency.

However I wonder if it wouldn't be better to make the generic parser parameterizeable, but still quite strict, so it only parses things that follow a fixed `<mnemonic> <op>, <op>, ...` structure for example and nothing else. For that, supplying a a more restricted parameter set, like one (optional) pattern per op might actually work. So kind of a mixture of `tokenize_asm_custom()` and `tokenize_asm_generic()` as it is now.
And the alternative if a plugin needs more freedom would be allowing them to do the entire parsing in C by themselves, without forcing them to use regex or anything else. The tokenized string could be returned as part of the regular `op` callback already.

> Sorry if I missed the docs about this. But what do you mean with API compatibility?

In short, for plugins, API compatibility is preserving the ability to compile the same plugin source against updated rizin. ABI compatibility is preserving the ability to use the same already compiled binary plugin inside updated rizin. For example extending the size of an array in the plugin struct may preserve API compatibility (mostly), but break ABI compatibility because the layout is moved around.

Reply:
How about this:

The generic method uses the register profile as parameter.
- If the current asm plugin provides no register profile, it parses the string like this: `<mnemmonic> <op>, <op>` where each non number `<op>` is treated as register.
- With register profile it still parses like `<mnemmonic> <op>, <op>`, but with less false positive register tokens. Every alphanumeric `<op>` which is no register is an unknown.

This way the generic method can be used right away without adding some kind of reduced parameter set for each arch.

Custom method:
- Do the parsing completely in the plugin if `op()` or `disassembly()` is called and save the `RzList*` in `RzAsmOp`/`RzAnalysisOp` (in the future maybe `RzArchOp`?).
- Coloring simply uses the generic method if the `RzList *` is NULL.
- This way @ret2libc remarks regarding keyword highlighting are addressed as well, because the plugin knows the analysis type.

Reply:
This sounds perfect to me. Parameters could still be added to the generic method, but only if they turn out to really be needed. If it works well enough from just the reg profile, then that is even better of course.